File: 43. Multiply Strings.java
class Solution {
    public String multiply(String num1, String num2) {
        int[] r = new int[num1.length() + num2.length()];
        for (int i = num1.length() - 1; i >= 0; i--) {
            for (int j = num2.length() - 1; j >= 0; j--) {
                int k = num1.length() - 1 - i + num2.length() - 1 - j;
                int t = Character.digit(num1.charAt(i), 10) * Character.digit(num2.charAt(j), 10) + r[k];
                r[k] = t % 10;
                r[k + 1] += t / 10;
            }
        }
        int i = r.length - 1;
        while (i >= 0 && r[i] == 0) {
            i--;
        }
        if (i == -1) {
            return "0";
        }
        StringBuilder sb = new StringBuilder();
        for (; i >= 0; i--) {
            sb.append(r[i]);
        }
        return sb.toString();
    }
}

File: 43. Multiply Strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        n1 = len(num1)
        n2 = len(num2)
        n = n1 + n2
        ret = [0] * n
        for index1, number1 in enumerate(reversed(num1)):
            for index2, number2 in enumerate(reversed(num2)):
                val = int(number1) * int(number2) + ret[n - (index1 + index2) - 1]
                add, mod = divmod(val, 10)
                ret[n - (index1 + index2) - 1] = mod
                advance = 1
                while add:
                    add += ret[n - (index1 + index2 + advance) - 1]
                    add, mod = divmod(add, 10)
                    ret[n - (index1 + index2 + advance) - 1] = mod
                    advance += 1

        left = 0
        while left < n and ret[left] == 0:
            left += 1
        return "".join(map(str, ret[left:])) if left < n else '0'